VertxApplication main should not print usage on application failure

When using `VertxApplication#main` and the verticle fails to start, the usage is printed after the error has been reported to the console, e.g.

```
Oct 19, 2024 1:52:42 PM io.vertx.launcher.application.VertxApplication
SEVERE: Failed in deploying verticle
io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:2528
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.Net.pollConnect(Native Method)
	at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:682)
	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:973)
	at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:330)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:384)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.handle(AbstractNioChannel.java:432)
	at io.netty.channel.nio.NioIoHandler$DefaultNioRegistration.handle(NioIoHandler.java:386)
	at io.netty.channel.nio.NioIoHandler.processSelectedKey(NioIoHandler.java:585)
	at io.netty.channel.nio.NioIoHandler.processSelectedKeysOptimized(NioIoHandler.java:560)
	at io.netty.channel.nio.NioIoHandler.processSelectedKeys(NioIoHandler.java:501)
	at io.netty.channel.nio.NioIoHandler.run(NioIoHandler.java:478)
	at io.netty.channel.SingleThreadIoEventLoop.runIo(SingleThreadIoEventLoop.java:184)
	at io.netty.channel.SingleThreadIoEventLoop.run(SingleThreadIoEventLoop.java:156)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:1123)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)

io.vertx.launcher.application.impl.CommandException
Usage: VertxApplication [-chw] [-vt] [-cluster-host=<clusterHost>]
                        [-cluster-port=<clusterPort>]
                        [-cluster-public-host=<clusterPublicHost>]
                        [-cluster-public-port=<clusterPublicPort>]
                        [-conf=<configStr>]
                        [-deployment-options=<deploymentOptionsStr>]
                        [-instances=<instances>] [-options=<vertxOptionsStr>]
                        <mainVerticle>
Runs a Vert.x application.
      <mainVerticle>        The main verticle fully qualified class name.
      -options, --options, -vertx-options, --vertx-options=<vertxOptionsStr>
                            Specifies the Vert.x options.
                            It should reference either a JSON file which
                              represents the options OR be a JSON string.
  -c, -cluster, --cluster   If specified, then the Vert.x instance will form a
                              cluster with any other Vert.x instances on the
                              network.
      -cluster-port, --cluster-port=<clusterPort>
                            Port to use for cluster communication.
                            By default, a spare random port is chosen.
      -cluster-host, --cluster-host=<clusterHost>
                            Host to bind to for cluster communication.
                            If this is not specified, Vert.x will attempt to
                              choose one from the available interfaces.
      -cluster-public-port, --cluster-public-port=<clusterPublicPort>
                            Public port to use for cluster communication.
                            By default, Vert.x uses the same as the cluster
                              port.
      -cluster-public-host, --cluster-public-host=<clusterPublicHost>
                            Public host to bind to for cluster communication.
                            By default, Vert.x uses the same as the cluster
                              host.
      -deployment-options, --deployment-options=<deploymentOptionsStr>
                            Specifies the main verticle deployment options.
  -w, -worker, --worker     If specified, then the main verticle is deployed
                              with the worker threading model.
                            Takes precedence over the value defined in
                              deployment options.
      -vt, -virtual-thread, --virtual-thread
                            If specified, then the main verticle is deployed
                              with the virtual thread threading model.
                            Takes precedence over the value defined in
                              deployment options.
      -instances, --instances=<instances>
                            Specifies how many instances of the verticle will
                              be deployed.
                            Takes precedence over the value defined in
                              deployment options.
      -conf, --conf=<configStr>
                            Specifies configuration that should be provided to
                              the verticle.
                            It should reference either a JSON file which
                              represents the options OR be a JSON string.
  -h, -help, --help         Display a help message.
```

--- a/application/src/main/java/io/vertx/launcher/application/ExitCodes.java
+++ b/application/src/main/java/io/vertx/launcher/application/ExitCodes.java
@@ -13,6 +13,8 @@ package io.vertx.launcher.application;
 
 public class ExitCodes {
 
+  public static final int SOFTWARE = 1;
+  public static final int USAGE = 2;
   public static final int VERTX_INITIALIZATION = 11;
   public static final int VERTX_DEPLOYMENT = 15;
 

--- a/application/src/main/java/io/vertx/launcher/application/VertxApplication.java
+++ b/application/src/main/java/io/vertx/launcher/application/VertxApplication.java
@@ -13,7 +13,7 @@ package io.vertx.launcher.application;
 
 import io.vertx.core.internal.logging.Logger;
 import io.vertx.core.internal.logging.LoggerFactory;
-import io.vertx.launcher.application.impl.CommandException;
+import io.vertx.launcher.application.impl.CommandExceptionHandler;
 import io.vertx.launcher.application.impl.VertxApplicationCommand;
 import picocli.CommandLine;
 
@@ -97,9 +97,9 @@ public class VertxApplication {
     VertxApplicationCommand command = new VertxApplicationCommand(this, Objects.requireNonNull(hooks), log);
     CommandLine commandLine = new CommandLine(command)
       .setOptionsCaseInsensitive(true)
-      .setExitCodeExceptionMapper(CommandException.EXIT_CODE_EXCEPTION_MAPPER);
+      .setExecutionExceptionHandler(new CommandExceptionHandler());
     int exitCode = commandLine.execute(args);
-    if (exitCode != 0) {
+    if (exitCode == ExitCodes.USAGE) {
       if (printUsageOnFailure) {
         CommandLine.usage(command, System.out, Ansi.ON);
       }

--- a/application/src/main/java/io/vertx/launcher/application/impl/CommandException.java
+++ b/application/src/main/java/io/vertx/launcher/application/impl/CommandException.java
@@ -12,22 +12,20 @@
 package io.vertx.launcher.application.impl;
 
 import io.vertx.core.VertxException;
-import picocli.CommandLine.IExitCodeExceptionMapper;
 
+/**
+ * An exception than can be thrown to interrupt the processing of {@link VertxApplicationCommand} and specify which exit code should be used.
+ */
 public class CommandException extends VertxException {
-
-  public static final IExitCodeExceptionMapper EXIT_CODE_EXCEPTION_MAPPER = t -> {
-    if (t instanceof CommandException) {
-      CommandException commandException = (CommandException) t;
-      return commandException.code;
-    }
-    return 1;
-  };
 
   private final int code;
 
   public CommandException(int code) {
     super((String) null, true);
     this.code = code;
+  }
+
+  public int getCode() {
+    return code;
   }
 }

--- a/application/src/main/java/io/vertx/launcher/application/impl/CommandExceptionHandler.java
+++ b/application/src/main/java/io/vertx/launcher/application/impl/CommandExceptionHandler.java
@@ -1,0 +1,19 @@
+package io.vertx.launcher.application.impl;
+
+import io.vertx.launcher.application.ExitCodes;
+import picocli.CommandLine;
+
+import static picocli.CommandLine.IExecutionExceptionHandler;
+import static picocli.CommandLine.ParseResult;
+
+public class CommandExceptionHandler implements IExecutionExceptionHandler {
+
+  @Override
+  public int handleExecutionException(Exception ex, CommandLine commandLine, ParseResult parseResult) {
+    if (ex instanceof CommandException) {
+      CommandException commandException = (CommandException) ex;
+      return commandException.getCode();
+    }
+    return ExitCodes.SOFTWARE;
+  }
+}

--- a/application/src/test/java/io/vertx/launcher/application/VertxApplicationTest.java
+++ b/application/src/test/java/io/vertx/launcher/application/VertxApplicationTest.java
@@ -18,8 +18,8 @@ import io.vertx.core.internal.VertxInternal;
 import io.vertx.core.json.JsonObject;
 import io.vertx.core.metrics.MetricsOptions;
 import io.vertx.core.spi.VertxServiceProvider;
-import io.vertx.test.fakecluster.FakeClusterManager;
 import io.vertx.test.TestVerticle;
+import io.vertx.test.fakecluster.FakeClusterManager;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
@@ -41,6 +41,7 @@ import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicReference;
 
 import static io.vertx.core.ThreadingModel.*;
+import static io.vertx.launcher.application.ExitCodes.USAGE;
 import static io.vertx.launcher.application.ExitCodes.VERTX_DEPLOYMENT;
 import static java.lang.Boolean.FALSE;
 import static java.lang.Boolean.TRUE;
@@ -119,13 +120,24 @@ public class VertxApplicationTest {
   }
 
   @Test
+  public void testUsageDisplayedWhenInputIsInvalid() throws Exception {
+    Integer exitCode = captureOutput(() -> {
+      TestVertxApplication app = new TestVertxApplication(new String[]{"-instances", "BOOM", HttpTestVerticle.class.getName()}, hooks);
+      return app.launch();
+    });
+    assertEquals(USAGE, exitCode);
+    assertTrue(out.toString().contains("Usage:"));
+    assertTrue(err.toString().contains("BOOM"));
+  }
+
+  @Test
   public void testFailureWhenBothWorkerAndVirtualOptionsAreSet() throws Exception {
     Integer exitCode = captureOutput(() -> {
       TestVertxApplication app = new TestVertxApplication(new String[]{"-vt", "-w", HttpTestVerticle.class.getName()}, hooks);
       return app.launch();
     });
     assertEquals(VERTX_DEPLOYMENT, exitCode);
-    assertTrue(out.toString().contains("Usage:"));
+    assertFalse(out.toString().contains("Usage:"));
     assertTrue(err.toString().contains("threading model"));
   }
 
@@ -145,7 +157,7 @@ public class VertxApplicationTest {
       return app.launch();
     });
     assertEquals(VERTX_DEPLOYMENT, exitCode);
-    assertTrue(out.toString().contains("Usage:"));
+    assertFalse(out.toString().contains("Usage:"));
   }
 
   @Test
@@ -156,7 +168,7 @@ public class VertxApplicationTest {
       return app.launch();
     });
     assertEquals(VERTX_DEPLOYMENT, exitCode);
-    assertTrue(out.toString().contains("Usage:"));
+    assertFalse(out.toString().contains("Usage:"));
     assertTrue(err.toString().contains("ClassNotFoundException"));
   }
 

